torch/csrc/autograd/python_variable.cpp

```diff
@@ -35,11 +41,13 @@ static PyObject* THPVariable_NewWithVar(PyTypeObject* type, Variable var)
   if (obj) {
     auto v = (THPVariable*) obj;
     new (&v->cdata) Variable(std::move(var));
-    v->cdata.get()->pyobj = obj;
-    if (auto fn = dynamic_cast<PyFunction*>(v->cdata.get()->_grad_fn.get())) {
+    v->cdata.set_pyobj(obj);
+    if (auto fn = dynamic_cast<PyFunction*>(v->cdata.grad_fn_ptr())) {
       // Create a new reference to the THPFunction. This ensures that ref count
       // of the THPFunction is at least the number of referring THPVariables.
-      v->cdata.get()->_grad_fn = THPFunction_asFunction((THPFunction*)fn->obj);
+      const auto output_nr = static_cast<uint32_t>(v->cdata.output_nr());
+      v->cdata.set_gradient_edge(
+          {THPFunction_asFunction((THPFunction*)fn->obj), output_nr});
```

Comment:
This looks semantically different, what's going on here?

Reply:
Yeah this one was a bit confusing. Previously we were only setting `grad_fn` and leaving the `output_nr` as it was. To be able to get rid of `set_grad_fn` in favor of `set_gradient_edge()`, I figured the equivalent would be to get the `output_nr` from before and set it as the new `output_nr` in the gradient_edge. So I believe the net-effect is the same? And I think `set_gradient_edge()` is the better API everywhere else since it seems less reasonable to change the underlying `grad_fn` without changing the `output_nr` -- we'd rather use the concept of an `Edge`.